### PR TITLE
[288] Add support for DnD of nodes inside React Flow diagrams

### LINKED ISSUE #288: Let end users move contained elements
Today, when an element is moved in its container, it is blocked at its borders.
The purpose is to allow the move of an element to another container and manage the semantic consequences.

It should thus be possible to move a child element from one container to another. This should work with free from layouted nodes and list layouted nodes. It should also allow moving border nodes from one node to another.

### PR DESCRIPTION
**Pull request by pcdavid**

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/dropNode/useDropNode.tsx [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    if (!loading && !error && data) {
+      const entries = data.viewer.editingContext.representation.description.dropNodeCompatibility;
+      setDropNodeCompatibilityData(entries);
+    }
+  }, [data, loading, error, setDropNodeCompatibilityData]);
+
+  const onDropNode = useDropNodeMutation();
+  const { getNodes, getIntersectingNodes } = useReactFlow();
+
+  const getNodeById: (string) => Node | null = (id: string) => getNodes().filter((n) => n.id === id)[0] || null;
```
Comment (frouene) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/dropNode/useDropNode.tsx]: Is it better to use find instead filter in this case?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/dropNode/useDropNode.tsx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+  const onNodeDragStart: NodeDragHandler = (_event, node) => {
+    setDraggedNode(node);
+    const dropDataEntry: GQLDropNodeCompatibilityEntry | undefined = dropNodeCompatibilityData.filter(
```
Comment (frouene) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/dropNode/useDropNode.tsx]: Same remark on the filter vs find

File: packages/view/backend/sirius-components-view-diagram-edit/src/main/java/org/eclipse/sirius/components/view/diagram/provider/DropNodeToolItemProvider.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-diagram-edit/src/main/java/org/eclipse/sirius/components/view/diagram/provider/DropNodeToolItemProvider.java]: 2023 only

File: packages/view/backend/sirius-components-view-diagram-edit/src/main/java/org/eclipse/sirius/components/view/diagram/provider/DropNodeToolItemProvider.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    /**
+     * This returns DropNodeTool.gif. <!-- begin-user-doc --> <!-- end-user-doc -->
+     *
+     * @generated
+     */
+    @Override
+    public Object getImage(Object object) {
+        return this.overlayImage(object, this.getResourceLocator().getImage("full/obj16/DropNodeTool"));
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-diagram-edit/src/main/java/org/eclipse/sirius/components/view/diagram/provider/DropNodeToolItemProvider.java]: The extension .svg is missing

File: packages/view/backend/sirius-components-view-diagram-edit/src/main/resources/icons/full/obj16/DropNodeTool.svg [hunk 5]
```diff
@@ -0,0 +1,70 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<svg
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-diagram-edit/src/main/resources/icons/full/obj16/DropNodeTool.svg]: Same icon as for drop tool, does it need to be duplicate ?

**Review by frouene — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx [hunk 6]
```diff
@@ -225,6 +229,13 @@ export const DiagramRenderer = ({ diagramRefreshedEventPayload, selection, setSe
     onDelete(event);
   };
 
+  const { onNodeDragStart, onNodeDrag, onNodeDragStop } = useDropNode();
+  const { dropData } = useContext<DropNodeContextValue>(DropNodeContext);
+  const theme = useTheme();
+
+  const diagramForbidden = dropData.draggedNodeId !== null && !dropData.droppableOnDiagram;
+  const backgroundColor = diagramForbidden ? theme.palette.action.disabledBackground : '#ffffff';
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: useDropNode should directly return the colors to use. It should leverage useTheme and the dropData

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx [hunk 7]
```diff
@@ -16,40 +16,53 @@ import { Theme, useTheme } from '@material-ui/core/styles';
...
   return imageNodeStyle;
 };
 
 export const ImageNode = memo(({ data, isConnectable, id, selected }: NodeProps<ImageNodeData>) => {
   const theme = useTheme();
+  const { dropData } = useContext<DropNodeContextValue>(DropNodeContext);
+  const isCompatibleDropTarget: boolean = dropData.compatibleNodeIds.includes(id);
+  const isSelectedDropTarget: boolean = id === dropData.targetNodeId;
+  const faded =
+    data.faded || (dropData.draggedNodeId !== null && dropData.draggedNodeId !== id && !isCompatibleDropTarget);
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: All this should be in the useDropNode hook to provide something like this:

```
const { nodeStyleProvider } = useDropNode()
const dropStyle: React.CssProperties = nodeStyleProvider.getStyle(id);
```

and then

```
style={...imageNodeStyle(theme, data.style, selected, data.faded), ...dropStyle}
```

I don't need any more impact than that on the various nodes (and the future custom nodes)

File: packages/view/backend/sirius-components-view-diagram-edit/src/main/java/org/eclipse/sirius/components/view/diagram/provider/DropNodeToolItemProvider.java (hunk 3 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-diagram-edit/src/main/java/org/eclipse/sirius/components/view/diagram/provider/DropNodeToolItemProvider.java]: Done.

File: packages/view/backend/sirius-components-view-diagram-edit/src/main/java/org/eclipse/sirius/components/view/diagram/provider/DropNodeToolItemProvider.java (hunk 4 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-diagram-edit/src/main/java/org/eclipse/sirius/components/view/diagram/provider/DropNodeToolItemProvider.java]: Done.

File: packages/view/backend/sirius-components-view-diagram-edit/src/main/resources/icons/full/obj16/DropNodeTool.svg (hunk 5 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-diagram-edit/src/main/resources/icons/full/obj16/DropNodeTool.svg]: Changed to use https://fonts.google.com/icons?selected=Material%20Icons%20Outlined%3Apan_tool%3A

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/dropNode/useDropNode.tsx (hunk 1 above)
Comment (pcdavid) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/dropNode/useDropNode.tsx]: Done.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/dropNode/useDropNode.tsx (hunk 2 above)
Comment (pcdavid) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/dropNode/useDropNode.tsx]: Done.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx (hunk 7 above)
Comment (pcdavid) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: Yes, the API needed was not clear yet, hence the draft status.
I'll move the "gory details" about the data structures manipulated into the hook and only expose a simple API for users.

Comment (pcdavid) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: Done.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx (hunk 6 above)
Comment (pcdavid) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: Mostly done, but I need more than a single color and less than a full `React.CSSProperties` to control both the actual background color and the color(s) of the grids. It still needs some work in this area.

**Review by frouene — APPROVED** [review]

**Pull request merged**